### ISSUE TITLE
[x86][windows] Asserts in `!argx->IsVarAddr() || (argx->gtType == TYP_I_IMPL)`

### ISSUE THREAD

**Issue by sandreenko** [dotnet/coreclr#22190]
`JIT_Performance._CodeQuality_Roslyn_CscBench` and  `managed_Compilation` started to fail with `!argx->IsVarAddr() || (argx->gtType == TYP_I_IMPL)`, for example: [jitStress2_jitstressRegs2](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_jitstress2_jitstressregs2/724/testReport/).

There are [29 changes](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_jitstress2_jitstressregs2/724/) in this commit, but I am pretty sure the issue was revealed by #22148.

cc @dotnet/jit-contrib

**Comment by CarolEidt** [comment]
>  I am pretty sure the issue was revealed by #22148 

That seems like a pretty safe bet, especially since JitStress really likes to arbitrarily retype things.

**Comment by erozenfeld** [comment]
I'll investigate.

**Comment by sandreenko** [comment]
Also it fails with JitMinOpts as well (in our desktop testing).

**Comment by erozenfeld** [comment]
Here is what happens here:

We start with this tree for the arg:
```
               [000040] ------------ arg2            +--*  ADDR      byref 
               [000039] ------------                 |  \--*  FIELD     int    _firstRowId
               [000037] L-----------                 |     \--*  ADDR      byref 
               [000038] ------------                 |        \--*  LCL_VAR   struct V04 tmp2     
```

Note the `L` on `[000037]`, it means that `GTF_ADDR_ONSTACK` is set on the node. That's the flag that tells  morph to retype the node ad `TYP_I_IMPL`.

The struct then gets promoted:

```
               [000040] ------------ arg2            +--*  ADDR      byref 
               [000039] ------------                 |  \--*  FIELD     int    _firstRowId
               [000037] L-----------                 |     \--*  ADDR      byref 
               [000038] ------------                 |        \--*  LCL_VAR   struct(P) V04 tmp2         
                                                     |        \--*    ref    V04._reader (offs=0x00) -> V06 tmp4         
                                                     |        \--*    int    V04._firstRowId (offs=0x04) -> V07 tmp5         
                                                     |        \--*    int    V04._lastRowId (offs=0x08) -> V08 tmp6  
```
and then LocalAddressVisitor transforms it to

```
               [000040] ------------ arg2            +--*  ADDR      byref 
               [000039] ------------                 |  \--*  LCL_VAR   int   (AX) V07 tmp5         
```
Note no `L` on `[000040]`, i.e., `GTF_ADDR_ONSTACK` is not set.  Since `GTF_ADDR_ONSTACK` is not set the type is not changed to `TYP_I_IMPL`.

Then stressClone kicks in and we clone 
```
     (  3,  3) [000040] ------------ arg2 on STK  +--*  ADDR      byref 
     (  3,  2) [000039] ----G--N----              |  \--*  LCL_VAR   int   (AX) V07 tmp5         
```
to
```
     (  3,  3) [000102] L---G------- arg2 on STK  +--*  ADDR      byref 
     (  3,  2) [000103] ----G--N----              |  \--*  LCL_VAR   int   (AX) V07 tmp5         
```

Note that `[000102] ` has `L`, i.e., `GTF_ADDR_ONSTACK` is set. That's because `gtNewOperNode` sets it for any GT_ADDR of GT_LCL_VAR.

So we end up with a node has `GTF_ADDR_ONSTACK` set but is of type TYP_BYREF instead of TYP_I_IMPL and the assert fires.

The fix I'm going to try is to teach LocalAddressVisitor to set `GTF_ADDR_ONSTACK` in this case. That should help avoid some unnecessary GC reporting.

**Comment by mikedn** [comment]
> The fix I'm going to try is to teach LocalAddressVisitor to set GTF_ADDR_ONSTACK in this case. That should help avoid some unnecessary GC reporting.

Perhaps we can make it change the type to `TYP_I_IMPL` directly?

**Comment by erozenfeld** [comment]
Yes, that should work too.

**Comment by erozenfeld** [comment]
But I think it's better to set 'GTF_ADDR_ONSTACK' so that it's recognized by 'IsVarAddr'.

**Comment by erozenfeld** [comment]
I have a fix, will open a PR shortly.

**Comment by erozenfeld** [comment]
Fixed by #22275.
